packages/cli/src/main.ts

```diff
@@ -220,6 +221,104 @@ async function main() {
 
         console.log('install successful. id:', response.data.id);
     }
+    else if (process.argv[2] === 'shell') {
+        console.log = () => {};
+
+        const ip = process.argv[3] || '127.0.0.1';
+        const login = await getOrDoLogin(ip);
+        const sdk = await connectScryptedClient({
+            baseUrl: `https://${ip}`,
+            pluginId: '@scrypted/core',
+            username: login.username,
+            password: login.token,
+            axiosConfig: {
+                httpsAgent,
+            }
+        });
+
+        const termSvc = await sdk.systemManager.getDeviceByName<DeviceProvider>("@scrypted/core").getDevice("terminalservice");
+        if (!termSvc) {
+            throw Error("@scrypted/core does not provide a Terminal Service");
+        }
+
+        const termSvcDirect = await sdk.connectRPCObject<StreamService>(termSvc);
+        const queue = createAsyncQueue<Buffer | string>();
+
+        if (process.stdin.isTTY) {
+            process.stdin.setRawMode(true);
+        } else {
+            process.stdin.end(() => queue.end());
+        }
+
+        const dim = { cols: process.stdout.columns, rows: process.stdout.rows };
+        queue.enqueue(JSON.stringify({ dim }));
+
+        let bufferedLength = 0;
+        const MAX_BUFFERED_LENGTH = 64000;
+        process.stdin.on('data', async data => {
+            bufferedLength += data.length;
+            const promise = queue.enqueue(data).then(() => bufferedLength -= data.length);
+            if (bufferedLength >= MAX_BUFFERED_LENGTH) {
+                process.stdin.pause();
+                await promise;
+                if (bufferedLength < MAX_BUFFERED_LENGTH)
+                    process.stdin.resume();
+            }
+        });
+
+        async function* generator() {
+            try {
+                while (true) {
+                    const buffers = queue.clear();
+                    if (buffers.length) {
+                        let buffersStart = 0;
+
+                        // this wonky loop-concat-yield is to batch together
```

Comment:
This could be simplified significantly with 2 queues. Control messages put the message in the control queue and an empty buffer in the buffer queue. Clear control message queue first, send that if a message exists.